crates/ty_python_semantic/src/types.rs

```diff
@@ -11068,6 +11106,78 @@ impl<'db> CallableType<'db> {
     }
 }
 
+/// Converting a type "into a callable" can possibly return a _union_ of callables. Eventually,
+/// when coercing that result to a single type, you'll get a `UnionType`. But this lets you handle
+/// that result as a list of `CallableType`s before merging them into a `UnionType` should that be
+/// helpful.
+#[derive(Clone, Debug, Eq, PartialEq, get_size2::GetSize, salsa::Update)]
+pub(crate) struct CallableTypes<'db>(SmallVec<[CallableType<'db>; 1]>);
+
+impl<'db> CallableTypes<'db> {
+    pub(crate) fn none() -> Self {
+        CallableTypes(smallvec![])
+    }
+
+    pub(crate) fn one(callable: CallableType<'db>) -> Self {
+        CallableTypes(smallvec![callable])
+    }
+
+    pub(crate) fn from_elements(callables: impl IntoIterator<Item = CallableType<'db>>) -> Self {
+        CallableTypes(callables.into_iter().collect())
+    }
+
+    pub(crate) fn exactly_one(self) -> Option<CallableType<'db>> {
+        self.0.into_iter().exactly_one().ok()
+    }
```

Comment:
the implementation of `Itertools::exactly_one` is pretty complicated because it has to handle arbitrary iterators. I think we can implement this method more simply:

```suggestion
    pub(crate) fn exactly_one(self) -> Option<CallableType<'db>> {
        match self.0.as_slice() {
            [single] => Some(*single),
            _ => None,
        }
    }
```